File: video_20_coinChange.java
import java.util.*;
class video_20_coinChange {
//method 1: memoization
    // public static int solve(int ind, int amnt, int nums[], int dp[][]){
    //     //base case
    //     if(amnt==0)   return 0;
    //     if(ind<0)   return (int)(1e9);
    //     if(dp[ind][amnt]!=-1)   return dp[ind][amnt];

    //     int nTake = solve(ind-1, amnt, nums, dp);
    //     int take = (int)(1e9);
    //     if(nums[ind]<=amnt)
    //     take = 1 + solve(ind, amnt-nums[ind], nums, dp);

    //     return dp[ind][amnt] = Math.min(take, nTake);
    // }
    // public static int MinCoin(int[] nums, int amount) {
    //     // Code here
    //     int n = nums.length;
    //     int dp[][] = new int[n][amount+1];
    //     for(int i = 0; i<n; i++)
    //         for(int j = 0; j<=amount; j++)
    //             dp[i][j] = -1;
        
    //     int ans = (int)(1e9);
    //     ans = solve(n-1, amount, nums, dp);
    //     if(ans!=1e9)
    //         return ans;
    //     return -1;
    // }

//method 2: tabulation
    // public static int MinCoin(int[] nums, int amount) {
    //     // Code here
    //     int n = nums.length;
    //     int dp[][] = new int[n][amount+1];

    //     for(int val = 0; val<=amount; val++)
    //         if(val%nums[0]==0)  dp[0][val] = val/nums[0];
    //         else dp[0][val] = (int)(1e9);


    //     for(int ind = 1; ind<n; ind++){
    //         for(int val = 0; val<=amount; val++){
    //             int nTake = dp[ind-1][val];
    //             int take = (int)(1e9);
    //             if(nums[ind]<=val)
    //                 take = 1+dp[ind][val-nums[ind]];
                
    //             dp[ind][val] = Math.min(take, nTake);
    //         }
    //     }
    //     if(dp[n-1][amount]!=1e9)    return dp[n-1][amount];
    //     return -1;
    // }
//method 3: max space optimization
    public static int MinCoin(int[] nums, int amount) {
        // Code here
        int n = nums.length;
        int prev[] = new int[amount+1];

        for(int val = 0; val<=amount; val++)
            if(val%nums[0]==0)  prev[val] = val/nums[0];
            else prev[val] = (int)(1e9);


        for(int ind = 1; ind<n; ind++){
            int curr[] = new int[amount+1];
            for(int val = 0; val<=amount; val++){
                int nTake = prev[val];
                int take = (int)(1e9);
                if(nums[ind]<=val)
                    take = 1+curr[val-nums[ind]];
                
                curr[val] = Math.min(take, nTake);
            }
            prev = curr.clone();
        }
        if(prev[amount]!=1e9)    return prev[amount];
        return -1;
    }

    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int nums[] = new int[n];
        for(int i = 0; i<n; i++)    nums[i] = sc.nextInt();
        int amount = sc.nextInt();
        System.out.print(MinCoin(nums, amount));    
    }
}

File: video_20_coinChange.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;

/****************************************** MEMOIZATION *********************************************** */
// int solve(int ind, vector<int> &coins, int target, vector<vector<int>> &dp){
//     if(ind == 0){
//         if(target%coins[ind] == 0) return target/coins[ind];
//         return 1e9;
//     }
//     if(dp[ind][target] != -1) return dp[ind][target];

//     int ntake = solve(ind-1, coins, target, dp);
//     int take = 1e9;
//     if(coins[ind]<=target)
//     take = 1+solve(ind, coins, target-coins[ind], dp);

//     return dp[ind][target] = min(take, ntake);
// }

// int minimumElements(vector<int> &num, int x)
// {
//     // Write your code here.  
//     int n = num.size(), ans;
//     vector<vector<int>> dp(n, vector<int> (x+1, -1)); 
//     ans = solve(n-1, num, x, dp);
//     if(ans>=1e9) return -1;
//     return ans;
// }

/***************************************** TABULATION ************************************************ */
// int minimumElements(vector<int> &nums, int target)
// {
//     // Write your code here.
//     int n = nums.size();
//     vector<vector<int>> dp(n, vector<int> (target+1, 0));

//     for(int t = 0; t<=target; t++){
//         if(t%nums[0] == 0) dp[0][t] = t/nums[0];
//         dp[0][t] = 1e9;
//     }
    
//     for(int ind = 1; ind<n; ind++){
//         for(int t = 1; t<=target; t++){
//             int ntake = dp[ind-1][t];
//             int take = 1e9;
//             if(nums[ind]<=t)
//                 take = 1+dp[ind][t-nums[ind]];
//             dp[ind][t] = min(take, ntake);
//         }
//     }
//     int ans = dp[n-1][target];
//     if(ans>=1e9) return -1;
//     return ans;
// }

/******************************************** MAIN ************************************************** */
// int main(){
//     int x, n;
//     cin>>x>>n;
//     vector<int> num(n);
//     for(int i = 0; i<n; i++)
//     cin>>num[i];
//     cout<<minimumElements(num, x);
// }